### DISCUSSION TITLE
A Chinese student aged 18 who wanna learn Python by himself

### DISCUSSION THREAD

**Discussion by jackzion47** [microsoft/ML-For-Beginners#675] [Q&A]
zero based，eager to learn
and im so appreciated for ur help
谢谢๑•́

**Comment by jackzion47** [comment]
> zero based，eager to learn
> and im so appreciated for ur help
> 谢谢๑•́

first，i want to konw how to start coding.
a book? a teaching video? 
go to school？
secondly，how to  enhance my coding skills and what is important for it
at last ，is there any recommend book or class？
with sincere thanks

**Comment by mbaranr** [comment]
Hey Jack!

Firstly, I wanted to mention that it's great news that you wanted to dive into the beautiful domain of programming. You mention that you want advice on how to start coding. I am currently a second year computer science student at uni, and my first year of studies was made for people of different coding backgrounds (including the ones who didn't know to program at all!). That being said, I can walk you through some of the topics that I covered this last year (which I believe were the most important) to give you an overview of the things you would most likely prioritize to learn first. This will give you a solid background from where you can expand as you'd like. 

What I would suggest is starting by picking a programming language of your choice. If you are interested in machine learning, I would go with python due to the plethora of tools available for free that make your life much easier in this field (as well as its simplicity and versatility). From there, I think it is useful to learn procedural programming and the basic syntax of the language of your choice. Nowadays there are plenty of ways to learn this for free through the internet that don't require paying for a univeristy degree. There is a particularly good course I would recommend, if it is python that you want to learn, provided by the university of Michigan. The name is somewhat along the lines of "Python for everybody". This is a wonderful introduction to the language that will definetely motivate you to continue learning. From there, you would need to take a look at algorithmic analysis and datastructures. This second basicaly means "How data is managed and stored". You would definetely need to start simple with learning different data types and then building up to structures like arrays, dictionaries, stacks, etc. After that, you can even explore more complex datastructues like graphs, binary trees, hashmaps, etc. From the experience I have had so far, I think that this topics are crucial for a smooth transition into the field (as data is stored everywhere!).

For now, learning this might take you some time so I don't want to overwhelm you with more topics. Actually, the same professor of the university of michigan offers you another course on datastructures which you might also like. Now, let's look at the second part of the question on how to enhance your coding skills. Even if it sounds cliche or boring, I must say the most important thing you need to improve is always practice. If you don't practice, it will be really hard to get a nice workflow when working on bigger projects. Some people say it has to due with your coding environment (text editor or IDE where you type the code), or that you may need a powerful computer to run your programs. However, eventhough this may play a small role, I believe the most important part is persevereance and a never ending sense of curiosity towards unkown problems. At the end of the day, most of programming is based on problem-solving, and a way to get better at this is by practicing on real-world problems. In that sense, I would suggest you to pick a project of your interest, break it into sub-problems or tasks, and don't hesitate to start working on it. The more you expose yourself to new problems, the more you will start to think in programming terms, the better you become.

That being said, I want to wish you the best of lucks in this new journey of yours and if you need more help don't hesitate to reach out to the community again!

**Reply by abdulqs** [reply]
Hello M4mbo. I came across your reply and found it so useful. I am also in the same page with Jack and wish to start a path on AI but want to start with ML. Please kindly guide me through. Thanks
